MediaSession should keep its action handlers alive

rreno opened the pull request:
#### 0743137c2f4f602963fc409ebc56ecd6690d204d
<pre>
MediaSession should keep its action handlers alive
<a href="https://bugs.webkit.org/show_bug.cgi?id=256334">https://bugs.webkit.org/show_bug.cgi?id=256334</a>
rdar://108913983

Reviewed by Chris Dumez.

The MediaSessionActionHandler callback is now a weak reference and so
when the MediaSession is visited it needs to mark any action handlers
that have been added to it.

* Source/WebCore/Modules/mediasession/MediaSession.cpp:
(WebCore::MediaSession::setActionHandler):
(WebCore::MediaSession::callActionHandler):
* Source/WebCore/Modules/mediasession/MediaSession.h:
(WebCore::MediaSession::hasActiveActionHandlers const):
(WebCore::MediaSession::visitActionHandlers const):
* Source/WebCore/Modules/mediasession/MediaSession.idl:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/JSMediaSessionCustom.cpp: Added.
(WebCore::JSMediaSession::visitAdditionalChildren):

Canonical link: <a href="https://commits.webkit.org/263715@main">https://commits.webkit.org/263715@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c811ccbb0d2fc7a5c568491fcde4584135de3bea

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/5522 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/5669 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/5858 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/7068 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/5531 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/5520 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/5894 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/5643 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/6971 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/5623 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/5683 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/4940 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/7103 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/3160 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/4957 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/12096 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/5026 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/5037 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/6783 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/5474 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/4485 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/4929 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1319 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/9033 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/5291 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/Modules/mediasession/MediaSession.h:
```diff
@@ -67,6 +67,8 @@ class MediaSession : public RefCounted<MediaSession>, public ActiveDOMObject, pu
 
     void callActionHandler(const MediaSessionActionDetails&, DOMPromiseDeferred<void>&&);
 
+    auto& actionHandlers() { return m_actionHandlers; }
```
should be marked as const. We don't want the callers to be able to modify the map.

rreno replied:
I took your advice below and wrote a function to iterate the map instead of this way of having the JSMediaSessionCustom code iterate the map.

cdumez commented on Source/WebCore/bindings/js/JSMediaSessionCustom.cpp:
```diff
@@ -0,0 +1,43 @@
+template <typename Visitor>
+void JSMediaSession::visitAdditionalChildren(Visitor& visitor)
+{
+    for (auto& actionHandler : wrapped().actionHandlers()) {
```
@rniwa @Constellation Do we need Lock for actionHandlers here? I believe visitAdditionalChildren() gets called off the main thread without synchronization so I suspect it isn't safe to iterate this map without locking but I am not 100% sure.

Constellation replied:
This function can be called concurrently while main thread is working. So if this map can be changed after the initialization, then it should be guarded by the lock.

rniwa replied:
Yeah, we need a lock.

cdumez replied:
The actionHandlers map seems to change dynamically so we need a lock. My proposal would be to add a `visitActionHandlers()` to MessageSession and have it grab a lock while iterating the map. Then call this visit function from here.

rreno replied:
I copied the pattern I saw for other HashMaps that have locks. Thanks!

cdumez commented on Source/WebCore/Modules/mediasession/MediaSession.h:
```diff
@@ -156,11 +159,28 @@ class MediaSession : public RefCounted<MediaSession>, public ActiveDOMObject, pu
+    return !m_actionHandlers.isEmpty();
+}
+
+template <typename Visitor>
```
A template function in the cpp even though the function is used from outside the cpp? I am a bit surprised this builds without explicit template instantiations. Maybe you're getting lucky because of unified builds?

cdumez replied:
I think you'd need explicit template instantiations for  `JSC::AbstractSlotVisitor` & `JSC::SlotVisitor`:
```
template void MediaSession:: visitActionHandlers(JSC::AbstractSlotVisitor&);
template void MediaSession:: visitActionHandlers(JSC::SlotVisitor&);
```

rreno replied:
I'm pretty sure this is the .h file. 

I'd be happy to use explicit instantiations and move it to the cpp though if you want.

cdumez replied:
Oh! Got confused by the diff format. You can keep as-is then. Don't know why I thought this was the cpp file :)

rreno replied:
I did originally try putting this in the cpp file without explicit instantiations so I got concerned I pushed a bad commit! I'll leave it alone then.

cdumez commented on Source/WebCore/Modules/mediasession/MediaSession.cpp:
```diff
@@ -268,6 +269,7 @@ void MediaSession::callActionHandler(const MediaSessionActionDetails& actionDeta
 
 bool MediaSession::callActionHandler(const MediaSessionActionDetails& actionDetails, TriggerGestureIndicator triggerGestureIndicator)
 {
+    Locker lock { m_actionHandlersLock };
```
We should consider reducing the scope of the lock. Something like:
```
RefPtr<MediaSessionActionHandler> handler;
{
    Locker lock { m_actionHandlersLock };
    handler = m_actionHandlers.get(actionDetails.action);
}
if (handler) {
```

rreno replied:
Yes! Will do.

cdumez commented on Source/WebCore/Modules/mediasession/MediaSession.cpp:
```diff
@@ -237,6 +237,7 @@ void MediaSession::setPlaybackState(MediaSessionPlaybackState state)
 
 void MediaSession::setActionHandler(MediaSessionAction action, RefPtr<MediaSessionActionHandler>&& handler)
 {
+    Locker lock { m_actionHandlersLock };
```
We should consider reducing the scope of the lock to just the HashMap operations.

cdumez commented on Source/WebCore/Modules/mediasession/MediaSession.cpp:
```diff
@@ -239,11 +239,13 @@ void MediaSession::setActionHandler(MediaSessionAction action, RefPtr<MediaSessi
 {
     if (handler) {
         ALWAYS_LOG(LOGIDENTIFIER, "adding ", action);
+        Locker lock { m_actionHandlersLock };
```
Still too wide a scope. We should only include the HashMap operation, not unrelated operation and external function calls.

cdumez commented on Source/WebCore/Modules/mediasession/MediaSession.cpp:
```diff
@@ -239,11 +239,13 @@ void MediaSession::setActionHandler(MediaSessionAction action, RefPtr<MediaSessi
         if (platformCommand != PlatformMediaSession::NoCommand)
             PlatformMediaSessionManager::sharedManager().addSupportedCommand(platformCommand);
     } else {
+        Locker lock { m_actionHandlersLock };
```
ditto

rreno replied:
Should I be concerned with the ALWAYS_LOG line in this scope? If I read correctly it could call a bunch of observers but it looks like the log channel isn't on by default.

rreno replied:
By concerned I mean should I release the lock before calling ALWAYS_LOG and then re-take it again before interacting with the map

cdumez replied:
I think this can be rewritten like so:
```
bool containedAction;
{ 
    Locker lock { m_actionHandlersLock };
    containedAction  = m_actionHandlers.remove(action)
}
if (containedAction)
    ALWAYS_LOG(LOGIDENTIFIER, "removing ", action);
```

Though it would probably be acceptable to include the ALWAYS_LOCK within the lock scope too.

cdumez commented on Source/WebCore/Modules/mediasession/MediaSession.cpp:
```diff
@@ -268,7 +270,12 @@ void MediaSession::callActionHandler(const MediaSessionActionDetails& actionDeta
 bool MediaSession::callActionHandler(const MediaSessionActionDetails& actionDetails, TriggerGestureIndicator triggerGestureIndicator)
 {
-    if (auto handler = m_actionHandlers.get(actionDetails.action)) {
+    RefPtr<MediaSessionActionHandler> handler;
```
Presumably, MediaSessionActionHandler needs to be ThreadSafeRefCounted now.

rreno replied:
I don't think this can be called off the main thread. Is it because the visit function can be called off the main thread so when iterating the map we increment/decrement the ref count?

cdumez replied:
You're right. No need for ThreadSafeRefCounted. The visit function is the one that gets called on an another thread and that one doesn't increment/decrement the ref count. My bad.

cdumez commented on Source/WebCore/Modules/mediasession/MediaSession.cpp:
```diff
@@ -239,14 +239,26 @@ void MediaSession::setActionHandler(MediaSessionAction action, RefPtr<MediaSessi
             PlatformMediaSessionManager::sharedManager().addSupportedCommand(platformCommand);
     } else {
-        if (m_actionHandlers.contains(action)) {
+        bool containsAction;
```
I think my proposal was better. The code was already unnecessarily inefficient by doing a double hash map look up and now is worse with double locking. Why didn’t you go with my proposal, was there something wrong with it?

rreno replied:
It's because I misread what you wrote above. I thought you were saying to use double locking but you were saying to use the return value of HashMap::remove

rreno replied:
I'll change it shortly.

rreno replied:
Ok I think I've tightened up the locks efficiently now. I'll keep this in mind in the future - keep an eye out for inefficient lookups and look for chances to keep a lock as tight as possible to the calls to the protected object.

cdumez reviewed (approved):
LGTM

rreno committed:
MediaSession should keep its action handlers alive
https://bugs.webkit.org/show_bug.cgi?id=256334
rdar://108913983

Reviewed by Chris Dumez.

The MediaSessionActionHandler callback is now a weak reference and so
when the MediaSession is visited it needs to mark any action handlers
that have been added to it.

* Source/WebCore/Modules/mediasession/MediaSession.cpp:
(WebCore::MediaSession::setActionHandler):
(WebCore::MediaSession::callActionHandler):
* Source/WebCore/Modules/mediasession/MediaSession.h:
(WebCore::MediaSession::hasActiveActionHandlers const):
(WebCore::MediaSession::visitActionHandlers const):
* Source/WebCore/Modules/mediasession/MediaSession.idl:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/JSMediaSessionCustom.cpp: Added.
(WebCore::JSMediaSession::visitAdditionalChildren):

Canonical link: https://commits.webkit.org/263715@main
